File: phyton/09-2024/structuraa/bugPlanning.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

class SIA {
public:
    class StaticStack {
    private:
        vector<int> data;
        int size;

    public:
        StaticStack(int capacity) : data(capacity, 0), size(0) {}

        void pop() {
            if (size == 0) {
                cout << "No element to delete" << endl;
            } else {
                --size;
            }
        }

        void add(int value) {
            if (size == data.size()) {
                cout << "List is full" << endl;
            } else {
                data[size++] = value;
            }
        }

        int getLast() const {
            if (size > 0) {
                return data[size - 1];
            } else {
                throw runtime_error("Stack is empty");
            }
        }

        void printList() const {
            for (int i = 0; i < size; ++i) {
                cout << data[i];
                if (i < size - 1) cout << " ";
            }
            cout << endl;
        }

        int getSize() const {
            return size;
        }
    };

    void gis(const vector<int>& data, int size) {
        StaticStack answer(size);
        answer.add(data[0]);

        for (size_t i = 1; i < data.size(); ++i) {
            if (data[i] > answer.getLast()) {
                answer.add(data[i]);
            }
        }

        cout << answer.getSize() << endl;
        answer.printList();
    }
};

int main() {
    SIA answer;
    int size;
    cin >> size;
    vector<int> data(size);

    for (int i = 0; i < size; ++i) {
        cin >> data[i];
    }

    answer.gis(data, size);

    return 0;
}

File: phyton/09-2024/structuraa/bugPlanning.py
class sia():
    class staticStack():
        def __init__(self, size: int):            
            self.data = [None]*size
            self.size = 0
        
        def pop(self):
            if(self.size == 0):
                print('Non element to delete')
            else: self.size -= 1
            
        def add(self, value: int):
            if(self.size == len(self.data)):
                print('list is full')
            else:
                self.data[self.size] = value
                self.size += 1
        
        def get_last(self):
            return self.data[self.size-1]
        
        def print_list(self):
            print("", end="")
            for i in range(self.size):
                print(self.data[i], end=" " if i < self.size - 1 else "")
            print("")

                
    def gis(self, data: str, size: int):
        answer = self.staticStack(size)
        answer.add(int(data[0]))
        for letter in data:
            if(letter == ' '):
                pass
            else:
                letter = int(letter) 
                if(letter > answer.get_last()):
                    answer.add(letter)

        print(answer.size)
        answer.print_list()
        
            
if '__main__' == __name__:
    answer = sia()
    size = input()
    data = input()
    answer.gis(data, int(size))
    